### PR TITLE
[Merged by Bors] - ET-3517 elastic search client in app state (7)

### PR DESCRIPTION
**Pull request by rustonaut**
Added an elastic search client to `AppState`.

**References:**

- [ET-3517]
- based on:
  - #671   
  - #670 
  - #669
  - #668
  - #667
  - #659

[ET-3517]: https://xainag.atlassian.net/browse/ET-3517?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: discovery_engine_core/web-api2/src/error/application.rs [hunk 1]
```diff
@@ -105,11 +105,11 @@ macro_rules! impl_application_error {
                 stringify!($name)
             }
 
-            fn encode_details(&self) -> serde_json::Value {
+            fn encode_details(&self) -> ::serde_json::Value {
                 ::serde_json::to_value(self)
                     .unwrap_or_else(|err| {
-                        error!(%err, "serializing error details failed");
-                        serde_json::Value::Null
+                        ::tracing::error!(%err, "serializing error details failed");
```
Comment (janpetschexain) [inline discovery_engine_core/web-api2/src/error/application.rs:111]: :wrench: the macro is already imported above, no need for a full path here
```suggestion
                        error!(%err, "serializing error details failed");
```

File: discovery_engine_core/web-api2/src/error/application.rs [hunk 2]
```diff
@@ -105,11 +105,11 @@ macro_rules! impl_application_error {
...
+                        ::serde_json::Value::Null
```
Comment (janpetschexain) [inline discovery_engine_core/web-api2/src/error/application.rs:112]: ❓ why the additional `::` qualifier? it's quite unlikely that we confuse the external crate with some internal path. even if we decide to take this precautionary measure, it then should be applied everywhere in the code, which should be done in a separate pr.

**Comment by biernacki** [comment]
please keep also in mind that our elastic client in `web-api` had some fixes introduced in #677 and #679

File: discovery_engine_core/web-api2/src/error/application.rs (hunk 1 above)
Comment (rustonaut) [reply discovery_engine_core/web-api2/src/error/application.rs:111]: The macro is used across the crate, so `error!` might not have been imported.

File: discovery_engine_core/web-api2/src/error/application.rs (hunk 2 above)
Comment (rustonaut) [reply discovery_engine_core/web-api2/src/error/application.rs:112]: Just general best practice for macros with `#[macro_export]`,  I don't think we should bother applying it everywhere in the code it doesn't add any value until it causes a problem, at which point we can change it. But there is no reason to not use it here.

File: discovery_engine_core/web-api2/src/error/application.rs (hunk 1 above)
Comment (janpetschexain) [reply discovery_engine_core/web-api2/src/error/application.rs:111]: i missed the `#[macro_export]` which was hidden behind github's folding, my bad.

**Comment by janpetschexain** [comment]
> please keep also in mind that our elastic client in `web-api` had some fixes introduced in #677 and #679

@rustonaut ☝️

**Comment by rustonaut** [comment]
I made a remainder ticket but this PR didn't include any of the affected code I only included the minimal skeleton.

The more specific code is in ET-3501/ET-3502.

**Review by janpetschexain — APPROVED** [review]

**Comment by rustonaut** [comment]
bors merge

**Comment by rustonaut** [comment]
bors merge
